### PR TITLE
give score for items removed from enemies

### PR DESCRIPTION
**Pull request by farooqkz**
This PR gives score to the players if they put an item dropped from teammate or enemy and put it in the teamchest. The item must not be stackable.

### CODE REVIEW

**Comment by AltZ-15** [comment]
What stops players from just cashing in all the items they've collected and talking it all back from the chest as soon as they get the points?

**Comment by farooqkz** [comment]
> What stops players from just cashing in all the items they've collected and talking it all back from the chest as soon as they get the points?

It's a reasonable concern. But why they would do that?

File: mods/ctf/ctf_teams/team_chest.lua [hunk 1]
```diff
@@ -4,6 +4,58 @@ local blacklist = {
...
+	["ctf_ranged:smg"] = 4,
+	["ctf_ranged:smg_loaded"] = 5,
+	["ctf_ranged:sniper_magnum"] = 8,
+	["ctf_ranged:sniper_magnum_loaded"] = 10,
+	["ctf_ranged:ammo"] = 4,
+	["default:diamond"] = 2.5,
+	["default:mese_crystal"] = 2,
+	["default:mese"] = 18,
+	["default:steel_ingot"] = 1,
+	["default:iron_lump"] = 1,
```
Comment (LoneWolfHT) [inline mods/ctf/ctf_teams/team_chest.lua:24]: Stackable

File: mods/ctf/ctf_teams/team_chest.lua [hunk 2]
```diff
@@ -4,6 +4,58 @@ local blacklist = {
...
+	["ctf_map:reinforced_cobble"] = 3,
+	["ctf_map:damage_cobble"] = 3,
+	["ctf_map:unwalkable_cobble"] = 1,
+	["ctf_map:unwalkable_stone"] = 1,
+	["ctf_map:unwalkable_dirt"] = 1,
+	["default:steelblock"] = 2.5,
+	["default:bronzeblock"] = 2.5,
+	["default:obsidian_block"] = 3.5,
+	["ctf_map:spike"] = 2.5,
+	["default:apple"] = 1.5,
```
Comment (LoneWolfHT) [inline mods/ctf/ctf_teams/team_chest.lua:49]: Stackable

File: mods/ctf/ctf_teams/team_chest.lua [hunk 3]
```diff
@@ -245,6 +297,15 @@ for _, team in ipairs(ctf_teams.teamlist) do
 			stack:to_string(),
 			minetest.pos_to_string(pos)
 		))
+		local meta = stack:get_meta()
+		if meta:get_string("dropped") == "true" then
+			local pname = player:get_player_name()
+			if pname then
+				local score = (item_value[stack:get_name()] or 0) * stack:get_count()
+				ctf_modebase:get_current_mode().recent_rankings.add(pname, { score = score }, false)
```
Comment (LoneWolfHT) [inline mods/ctf/ctf_teams/team_chest.lua:305]: Make sure this is crash-safe, there are times where there isn't an ongoing mode and `ctf_modebase:get_current_mode()` returns false/nil

File: mods/ctf/ctf_teams/team_chest.lua [hunk 4]
```diff
@@ -245,6 +297,15 @@ for _, team in ipairs(ctf_teams.teamlist) do
...
+			end
+			meta:set_string("dropped", "")
```
Comment (LoneWolfHT) [inline mods/ctf/ctf_teams/team_chest.lua:307]: Make sure this change is actually getting applied to the stack you put in.
I'm not sure if `stack` is a deep copy of the actual stack or not.

File: mods/pvp/dropondie/init.lua [hunk 5]
```diff
@@ -2,6 +2,10 @@ dropondie = {}
 
 local function drop_list(pos, inv, list)
 	for _, item in ipairs(inv:get_list(list)) do
+		if minetest.registered_items[item:get_name()].stack_max == 1 then
```
Comment (LoneWolfHT) [inline mods/pvp/dropondie/init.lua:5]: I'd be slightly worried about `stack_max` being nil/nonexistent here

**Review by LoneWolfHT — COMMENTED** [review]
You could replace the meta string `"dropped"` with an int value. `meta:get_int()` returns 0 for unset values, and you could treat 1 as truthy
`meta:set_int("dropped", 1)`, etc

**Comment by LoneWolfHT** [comment]
To play it safe I would prefer to only give 1 or less score for items we reward returning, and slowly increase that as we watch things play out on the main server

File: mods/ctf/ctf_teams/team_chest.lua (hunk 1 above)
Comment (farooqkz) [reply mods/ctf/ctf_teams/team_chest.lua:24]: I have assigned values to items for further use. The item metadata is added to an item only if it's not stackable.

File: mods/ctf/ctf_teams/team_chest.lua (hunk 4 above)
Comment (farooqkz) [reply mods/ctf/ctf_teams/team_chest.lua]: You are right. The metadata is not set so score is given if the player takes and puts over again and again

File: mods/pvp/dropondie/init.lua (hunk 5 above)
Comment (farooqkz) [reply mods/pvp/dropondie/init.lua:5]: I'll do a research about that.

**Comment by farooqkz** [comment]
> To play it safe I would prefer to only give 1 or less score for items we reward returning, and slowly increase that as we watch things play out on the main server

I disagree with the reward being the same, one point, for all the items. And less than one is a bad idea IMO

File: mods/pvp/dropondie/init.lua (hunk 5 above)
Comment (farooqkz) [reply mods/pvp/dropondie/init.lua:5]: On second thought, why should it matter if `stack_max` is `nil`?

Comment (farooqkz) [reply mods/pvp/dropondie/init.lua:5]: image

**Comment by LoneWolfHT** [comment]
> I disagree with the reward being the same, one point, for all the items. And less than one is a bad idea IMO

Did you read the `and slowly increase that as we watch things play out on the main server` part?

**Comment by farooqkz** [comment]
> > I disagree with the reward being the same, one point, for all the items. And less than one is a bad idea IMO
> 
> Did you read the `and slowly increase that as we watch things play out on the main server` part? The item values can be differentiated again as we increase them

Yeah I know but it doesn't make sense to me even temporary. But if you mean some stuff can be temporary non-sensible, that's your decision. I am committing to give exactly one for all items.

File: mods/pvp/dropondie/init.lua (hunk 5 above)
Comment (LoneWolfHT) [reply mods/pvp/dropondie/init.lua:5]: :+1:

Comment (LoneWolfHT) [reply mods/pvp/dropondie/init.lua:5]: > On second thought, why should it matter if `stack_max` is `nil`?

Comparing numbers with nil causes a crash

File: mods/ctf/ctf_teams/team_chest.lua (hunk 4 above)
Comment (LoneWolfHT) [reply mods/ctf/ctf_teams/team_chest.lua]: You can fix this by setting the stack in the destination index to your stack variable

**Comment by LoneWolfHT** [comment]
Thanks.
https://github.com/MT-CTF/capturetheflag/pull/1173#discussion_r1254898943 needs to be fixed and this will be ready for testing

File: mods/pvp/dropondie/init.lua (hunk 5 above)
Comment (farooqkz) [reply mods/pvp/dropondie/init.lua:5]: you sure?

```
13:33:49 ~ $ lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
> local a = nil
> if a == 1 then print("H") end
> if b == 1 then print("H") end
> 
```

Comment (LoneWolfHT) [reply mods/pvp/dropondie/init.lua:5]: Ah, `==` doesn't count as a comparison (unlike `>=`, etc) it seems. Good to know.

**Comment by LoneWolfHT** [comment]
Your rebase didn't work at all there

**Comment by LoneWolfHT** [comment]
With rebasing I tend to use `git rebase -i <PR branch>~<number of previous commits I want to have the option to squash>` (e.g `git rebase -i master~6`), squash all my commits together, then `git pull -r mainrepo master` to rebase off of the CTF repo's master branch

I'm laying that out to show you some things you can do, I highly recommend googling/figuring out what those commands do on your own though

**Comment by LoneWolfHT** [comment]
Items returned to team chest shouldn't give score if they're your own

**Comment by farooqkz** [comment]
@LoneWolfHT Done.

**Review by LoneWolfHT — APPROVED** [review]
Rebase conflict

**Comment by farooqkz** [comment]
> Rebase conflict

WDYM?

**Comment by LoneWolfHT** [comment]
image

**Comment by farooqkz** [comment]
@LoneWolfHT Done

**Comment by LoneWolfHT** [comment]
How are you rebasing your PR? This shouldn't be showing up: 
image

**Comment by farooqkz** [comment]
@LoneWolfHT What was state of this? Do I need to do anything?

**Pull request merged**